### PR TITLE
server-side and client-side advises for configuring smart card auth

### PR DESCRIPTION
**Pull request by martbab**
These advises generate scripts which configure all the components
required for successful processing and performing of smart card auth requests on IPA server and enrolled clients.

https://pagure.io/freeipa/issue/6982

### CODE REVIEW

**Comment by abbra** [comment]
It looks good, thanks. One small comment is that we probably want to allow the script to work on non-RPM platforms as well (Debian or Ubuntu too) where `yum` does not exist. On the other hand, we also would probably want to eventually allow setting this up against `mod_ssl` too. These two changes can be added later.

**Comment by martbab** [comment]
support for non-RPM platforms would require some more additions to the base Advice code to handle this systematically, or alternatively we may just test for the presence of required command and fail with instruction to install the missing package using platform-specific means.

While we may want to migrate to `mod_ssl` in the future, there will be much more work to do regarding switching TLS modules so the advise can be ported as a part of this effort.

**Comment by abbra** [comment]
Yes, I'm not asking you to implement these improvements as part of this PR. That's fine to be done in a separate effort.

**Comment by flo-renaud** [comment]
Hi @martbab 
thank you for the patch. Works great!
We could also enhance the script:
- check that it is run by the root user
- to enable NSSOCP, the regex does not match "##NSSOCSP off" but should. 
- the script must be run on each IPA server, maybe we should make it clearer.
- the script could also configure /etc/sssd.conf with
[pam]
pam_cert_auth = True

This could be done in a later commit, I can open a RFE if needed

**Comment by martbab** [comment]
@flo thanks for your input, I will rework the PR tomorrow.

**Comment by martbab** [comment]
@flo @abbra I have rebased PR and included also a recipe for client configuration for the sake of completeness.

**Comment by abbra** [comment]
Thanks. Comments so far:
* client configuration does not make sure to ask for a removal of `pam_pkcs11` package
* client configuration does not call for the `authconfig` to configure PAM stack. It seems to me that we either need a second recipe that enables IPA client system for a remote logon or need to integrate it in the same recipe
* there are few places in the server plugin that do an explicit 'if ... ' command. This probably would be better to replace with `commands_on_predicate` like in the client case

**Comment by martbab** [comment]
@abbra thanks for review. Is `pam_pkcs11` removal necessary for client? Also what option does the recipe need to pass to `authconfig` to properly configure smart card auth? isn't it enough to configure SSSD?

**Comment by abbra** [comment]
It is all documented in https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7-Beta/html/Linux_Domain_Identity_Authentication_and_Policy_Guide/smart-cards.html (notice 7-Beta, this is a new material for RHEL 7.4 beta), section 22.5.1.3.

**Comment by martbab** [comment]
That section[1] only instructs to configure `pam_cert_auth=true` in the SSSD's `pam` section which is already done on both server and client, see `enable_pam_auth_in_sssd` method. Am I missing something? 

[1] https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7-Beta/html/Linux_Domain_Identity_Authentication_and_Policy_Guide/idm-smart-cards.html

**Comment by flo-renaud** [comment]
Hi @martbab 
I think @abbra was referring to this [section](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7-Beta/html/Linux_Domain_Identity_Authentication_and_Policy_Guide/smart-cards-advanced.html#sc-clients-config) running authconfig --enablesmartcard --smartcardmodule=sssd --updateall
This command will update sssd.conf with pam_cert_auth=True.

**Comment by martbab** [comment]
@flo ah sorry I missed that. I will incorporate it into advise then.

**Comment by martbab** [comment]
@flo regarding enabling Smart Card login ( add PKCS#11 module, configure SSSD and such), do we really need to setup this on server? I do not expect somebody logging directly to machine hosting FreeIPA server using smard card reader.

**Comment by abbra** [comment]
@martbab, this actually makes full sense -- if you want to increase the security of your IPA masters, you might force using smart cards only to do ssh login. We are not there yet by default but I can see it being a mandate for organizations that enforce industry or governmental requirements.

**Comment by abbra** [comment]
Note that "directly" may actually mean using a virtualized remote smart card access which is provided via virtualized USB pass-through done by your virtualization environment.

**Comment by martbab** [comment]
Also I get the following error when running authconfig:

```console
authconfig: Authentication module /lib64/security/pam_pkcs11.so is missing. Authentication process might not work correctly.
```

It is understandable, since I have removed pam_pkcs11 package as per documentation, but it still puzzles me.

It may be that I have an old version of authconfig, as I am developing this on F25 where I have authconfig-6.2.10-14.fc25.x86_64.

**Comment by abbra** [comment]
@martbab, definitely `authconfig` in fc25 is too old for this. On F26 I have version 7.0.1-1. It does announce support for SSSD smartcard enablement.

File: ipaserver/advise/plugins/smart_card_auth.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            ['Failed to install bind-utils'])
+
+        self.log.comment('make sure ipa-ca records are resolvable, '
+                         'otherwise error out and instruct')
+        self.log.comment('the user to update the DNS infrastructure')
+        self.log.command('ipaca_records=$(dig +short '
+                         'ipa-ca.{})'.format(ipa_domain_name))
+
+        self.log.exit_on_predicate(
+            '[ -z "ipaca_records" ]',
```
Comment (flo-renaud) [inline ipaserver/advise/plugins/smart_card_auth.py]: This should be $ipaca_records (the $ is missing).

Comment (martbab) [inline ipaserver/advise/plugins/smart_card_auth.py]: Oops sorry, should be fixed now.

**Comment by flo-renaud** [comment]
Hi @martbab 
Thank you for the fix. Works as expected.

**Comment by martbab** [comment]
master:

* 0569c02f17f853d97280f52f4a7fefecc72cf45d Extend the advice printing code by some useful abstractions
* e418e9a4ca747886c53d05ae80597834f1d3d021 Prepare advise plugin for smart card auth configuration


ipa-4-5:

* 7ea7ee4326679c098d3e4e4d6a2bc743707708ca Extend the advice printing code by some useful abstractions
* 84ca9761bd47f28b72581d1fe6bd8cfa824b6df3 Prepare advise plugin for smart card auth configuration

**Pull request closed without merging**